pkg/clients/redshift/redshift.go

```diff
@@ -0,0 +1,504 @@
+		AutomatedSnapshotRetentionPeriod: p.AutomatedSnapshotRetentionPeriod,
+		AvailabilityZone:                 p.AvailabilityZone,
+		ClusterIdentifier:                cid,
+		ClusterParameterGroupName:        p.ClusterParameterGroupName,
+		ClusterSecurityGroups:            p.ClusterSecurityGroups,
+		ClusterSubnetGroupName:           p.ClusterSubnetGroupName,
+		ClusterType:                      p.ClusterType,
+		ClusterVersion:                   p.ClusterVersion,
+		DBName:                           p.DBName,
+		ElasticIp:                        p.ElasticIP,
+		Encrypted:                        p.Encrypted,
+		EnhancedVpcRouting:               p.EnhancedVPCRouting,
+		HsmClientCertificateIdentifier:   p.HSMClientCertificateIdentifier,
+		HsmConfigurationIdentifier:       p.HSMConfigurationIdentifier,
+		IamRoles:                         p.IAMRoles,
+		KmsKeyId:                         p.KMSKeyID,
+		MaintenanceTrackName:             p.MaintenanceTrackName,
+		ManualSnapshotRetentionPeriod:    p.ManualSnapshotRetentionPeriod,
+		MasterUserPassword:               pw,
+		MasterUsername:                   p.MasterUsername,
+		NodeType:                         p.NodeType,
+		NumberOfNodes:                    p.NumberOfNodes,
+		Port:                             p.Port,
+		PreferredMaintenanceWindow:       p.PreferredMaintenanceWindow,
+		PubliclyAccessible:               p.PubliclyAccessible,
+		SnapshotScheduleIdentifier:       p.SnapshotScheduleIdentifier,
+		Tags:                             tags,
+		VpcSecurityGroupIds:              p.VPCSecurityGroupIDs,
+	}
+}
+
+// GenerateModifyClusterInput from RedshiftSpec
+func GenerateModifyClusterInput(p *v1alpha1.RedshiftParameters, cl redshift.Cluster) *redshift.ModifyClusterInput { //nolint:gocyclo
+	patch, err := CreatePatch(p, &cl)
+	if err != nil {
+		return &redshift.ModifyClusterInput{}
+	}
+
+	o := &redshift.ModifyClusterInput{ClusterIdentifier: cl.ClusterIdentifier}
+	if patch.AllowVersionUpgrade != nil {
+		o.AllowVersionUpgrade = p.AllowVersionUpgrade
+	}
+	if patch.AutomatedSnapshotRetentionPeriod != nil {
+		o.AutomatedSnapshotRetentionPeriod = p.AutomatedSnapshotRetentionPeriod
+	}
+	// If the cluster type, node type, or number of nodes changed, then the AWS API expects all three
+	// items to be sent over
+	// When a resize operation is requested, no other modifications are allowed in the same request
+	if patch.ClusterType != nil || patch.NodeType != nil || patch.NumberOfNodes != nil {
+		o.ClusterType = p.ClusterType
+		o.NodeType = p.NodeType
+		o.NumberOfNodes = p.NumberOfNodes
+		return o
+	}
+	if patch.ClusterVersion != nil {
+		o.ClusterVersion = p.ClusterVersion
+	}
+	// The request to modify publicly accessible options for the cluster cannot be made in the same request as other modifications
+	if patch.ElasticIP != nil || patch.PubliclyAccessible != nil {
+		o.ElasticIp = p.ElasticIP
+		o.PubliclyAccessible = p.PubliclyAccessible
+		return o
```

Comment:
As per the comment the `ModifyClusterInput` should just have `ElasticIp` and `PubliclyAccessible` populated at this state, but I think `AllowVersionUpgrade` and `AutomatedSnapshotRetentionPeriod` may also be set from the above lines.

@arush-sal Can this be the case ever?

Reply:
@sahil-lakhwani I have tested this and as the comment mention `"The request to modify publicly accessible options for the cluster cannot be made in the same request as other modifications"`. So either it can be a change to `ElasticIp` or `PubliclyAccessible` field and no other change or else there will be an error returned.